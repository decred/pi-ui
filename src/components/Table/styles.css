.tableWrapper {
  padding-bottom: 2rem;
}

.table {
  border-spacing: 0;
  border-collapse: collapse;
  width: 100%;
}

.pages {
  margin-top: 3rem;
  display: flex;
  align-items: center;
  justify-content: flex-end;
}

.pageItem {
  cursor: pointer;
  padding: 0.2rem 0.5rem;
  border-radius: 0.3rem;
  margin: 0 0.3rem;
  user-select: none;
}

.pageItemArrow {
  cursor: pointer;
  user-select: none;
}

.pageItemArrowDisabled {
  composes: pageItemArrow;
  cursor: not-allowed;
}

.pageItemActive {
  background-color: var(--color-gray-lighter);
  color: var(--color-primary);
  composes: pageItem;
}

.tableCell {
  max-width: 12rem;
  min-width: 12rem;
  overflow: hidden;
  text-align: center;
  text-overflow: ellipsis;
}

.tableHeadCell {
  composes: tableCell;
  height: 5.2rem;
}

.tableBodyCell {
  composes: tableCell;
  height: 5rem;
  border-bottom: 0.1rem solid var(--color-gray-lightest);
}

.tableHead {
  background-color: var(--color-gray-lightest);
}

@media screen and (max-width: 768px) {
  .table {
    border: 0;
  }

  .tableHead {
    border: none;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
  }

  .tableRow {
    border: 0.1rem solid #f3f5f6;
    padding: 1.3rem 2rem;
    display: block;
    margin-bottom: 1rem;
  }

  .tableCell {
    max-width: 100%;
  }

  .tableBodyCellText {
    overflow: hidden;
    max-width: 20rem;
    text-overflow: ellipsis;
  }

  .tableBodyCell {
    display: block;
    display: flex;
    align-items: center;
    border: 0;
    justify-content: space-between;
    font-size: 1.35rem;
  }

  td.tableBodyCell:not(:last-child) {
    border-bottom: 0.1rem solid #f3f5f6;
  }

  .tableBodyCell::before {
    content: attr(data-label);
    font-weight: bold;
  }
}
