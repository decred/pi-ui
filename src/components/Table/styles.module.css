.tableWrapper {
  padding-bottom: 2rem;
}

.table {
  table-layout: fixed;
  border-spacing: 0;
  border-collapse: collapse;
  width: 100%;
}

.paginator {
  margin-top: 3rem;
  justify-content: flex-end;
}

.tableCell {
  max-width: 12rem;
  min-width: 12rem;
  overflow-x: hidden;
  text-align: center;
  text-overflow: ellipsis;
}

.tableHeadCell {
  composes: tableCell;
  height: 5.2rem;
}

.tableBodyCell {
  composes: tableCell;
  height: 5rem;
  border-bottom: 0.1rem solid var(--table-cell-border-color);
}

.tableHead {
  background-color: var(--table-header-background);
}

@media (--sm-viewport) {
  .table {
    border: 0;
  }

  .tableHead {
    border: none;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
  }

  .tableRow {
    border: 0.1rem solid #f3f5f6;
    padding: 1.3rem 2rem;
    display: block;
    margin-bottom: 1rem;
  }

  .tableCell {
    max-width: 100%;
  }

  .tableBodyCellText {
    overflow-x: hidden;
    max-width: 20rem;
    text-overflow: ellipsis;
  }

  .tableBodyCell {
    display: flex;
    align-items: center;
    border: 0;
    justify-content: space-between;
    font-size: 1.35rem;
  }

  td.tableBodyCell:not(:last-child) {
    border-bottom: 0.1rem solid #f3f5f6;
  }

  .tableBodyCell::before {
    content: attr(data-label);
    font-weight: bold;
  }
}
