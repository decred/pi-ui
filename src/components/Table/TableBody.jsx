import React from "react";
import PropTypes from "prop-types";
import styles from "./styles.css";
import { classNames } from "../../utils";

const TableBody = ({ data, className, rowClassName, bodyCellClassName }) => {
  return (
    <tbody className={className}>
      {data.map((line, lineIdx) => (
        <tr
          key={`line-${lineIdx}`}
          className={classNames(styles.tableRow, rowClassName)}>
          {Object.keys(line).map((key, keyIdx) => (
            <td
              key={`field-${keyIdx}`}
              className={classNames(styles.tableBodyCell, bodyCellClassName)}
              data-label={key}>
              {typeof line[key] === "string" ? (
                <span className={styles.tableBodyCellText}>{line[key]}</span>
              ) : (
                line[key]
              )}
            </td>
          ))}
        </tr>
      ))}
    </tbody>
  );
};

TableBody.propTypes = {
  data: PropTypes.array.isRequired,
  className: PropTypes.string,
  rowClassName: PropTypes.string,
  bodyCellClassName: PropTypes.string
};

export default TableBody;
