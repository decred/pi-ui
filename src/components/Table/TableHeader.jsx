import React from "react";
import PropTypes from "prop-types";
import styles from "./styles.css";
import { classNames } from "../../utils";

const TableHeader = ({
  headers,
  headerCellClassName,
  rowClassName,
  className,
}) => {
  return (
    <thead className={classNames(styles.tableHead, className)}>
      <tr className={classNames(styles.tableRow, rowClassName)}>
        {headers.map((header, idx) => (
          <th
            key={`header-${idx}`}
            className={classNames(styles.tableHeadCell, headerCellClassName)}
            scope="col">
            {header}
          </th>
        ))}
      </tr>
    </thead>
  );
};

TableHeader.propTypes = {
  headers: PropTypes.arrayOf(PropTypes.node).isRequired,
  headerCellClassName: PropTypes.string,
  rowClassName: PropTypes.string,
  className: PropTypes.string,
};

export default TableHeader;
