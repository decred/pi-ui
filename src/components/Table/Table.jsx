import React, { useState } from "react";
import PropTypes from "prop-types";
import TableHeader from "./TableHeader.jsx";
import TableBody from "./TableBody.jsx";
import Icon from "../Icon/Icon.jsx";
import styles from "./styles.css";
import { classNames } from "../../utils";
import { useTheme, getThemeProperty } from "../../theme";

const Table = ({
  data,
  headers,
  linesPerPage,
  disablePagination,
  className,
  wrapperClassName,
  headClassName,
  headerCellClassName,
  rowClassName,
  bodyClassName,
  bodyCellClassName
}) => {
  if (linesPerPage < 1) {
    throw new Error("Invalid prop. linesPerPage should be bigger than 1");
  }
  const totalPages = Math.ceil(data.length / linesPerPage);
  const pagesArr = [...new Array(totalPages)].map((_, i) => i + 1);
  const [page, setPage] = useState(1);
  const startIndex = (page - 1) * linesPerPage;
  const canGoBack = page > 1;
  const canGoNext = page < totalPages;
  const previousPage = () => (canGoBack ? setPage(page - 1) : null);
  const nextPage = () => (canGoNext ? setPage(page + 1) : null);

  const { theme } = useTheme();
  const arrowIconBackground = getThemeProperty(
    theme,
    "table-page-arrow-background"
  );
  const disabledArrowColor = getThemeProperty(theme, "color-gray");
  const arrowColor = getThemeProperty(theme, "color-primary");
  return (
    <div className={classNames(styles.tableWrapper, wrapperClassName)}>
      <table className={classNames(styles.table, className)}>
        <TableHeader
          headers={headers}
          className={headClassName}
          headerCellClassName={headerCellClassName}
          rowClassName={rowClassName}
        />
        <TableBody
          data={data.slice(startIndex, startIndex + linesPerPage)}
          className={bodyClassName}
          rowClassName={rowClassName}
          bodyCellClassName={bodyCellClassName}
        />
      </table>
      {!disablePagination && pagesArr.length > 1 && (
        <div className={styles.pages}>
          <span
            className={
              canGoBack ? styles.pageItemArrow : styles.pageItemArrowDisabled
            }
            onClick={previousPage}>
            <Icon
              type="left"
              size="lg"
              backgroundColor={arrowIconBackground}
              iconColor={canGoBack ? arrowColor : disabledArrowColor}
            />
          </span>
          {pagesArr.map((item, idx) => (
            <span
              key={`page-number-${idx}`}
              className={
                item === page ? styles.pageItemActive : styles.pageItem
              }
              onClick={() => setPage(item)}>
              {item}
            </span>
          ))}
          <span
            className={
              canGoNext ? styles.pageItemArrow : styles.pageItemArrowDisabled
            }
            onClick={nextPage}>
            <Icon
              type="right"
              size="lg"
              backgroundColor={arrowIconBackground}
              iconColor={canGoNext ? arrowColor : disabledArrowColor}
            />
          </span>
        </div>
      )}
    </div>
  );
};

Table.propTypes = {
  data: PropTypes.array.isRequired,
  headers: PropTypes.arrayOf(PropTypes.string).isRequired,
  linesPerPage: PropTypes.number,
  disablePagination: PropTypes.bool,
  className: PropTypes.string,
  wrapperClassName: PropTypes.string,
  headClassName: PropTypes.string,
  headerCellClassName: PropTypes.string,
  rowClassName: PropTypes.string,
  bodyClassName: PropTypes.string,
  bodyCellClassName: PropTypes.string
};

Table.defaultProps = {
  linesPerPage: 10,
  disablePagination: false
};

export default Table;
