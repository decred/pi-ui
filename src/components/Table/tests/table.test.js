import React from "react";
import { create } from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
import Table from "../Table";
import {
  defaultLightTheme,
  ThemeProvider,
  DEFAULT_LIGHT_THEME_NAME
} from "../../../theme";

const mockData = [
  {
    Test1: "testl1c1",
    Test2: "testl1c2"
  },
  {
    Test1: "testl2c1",
    Test2: "testl2c2"
  }
];

const mockHeaders = Object.keys(mockData[0]);

describe("Table Component", () => {
  test("Matches the snapshot", () => {
    const table = create(
      <ThemeProvider
        themes={{ [DEFAULT_LIGHT_THEME_NAME]: defaultLightTheme }}
        defaultThemeName={DEFAULT_LIGHT_THEME_NAME}>
        <Table data={mockData} headers={mockHeaders} />
      </ThemeProvider>
    );
    expect(table.toJSON()).toMatchSnapshot();
  });

  test("Table pagination", () => {
    const { getByTestId, queryByText, queryByTestId } = render(
      <ThemeProvider
        themes={{ [DEFAULT_LIGHT_THEME_NAME]: defaultLightTheme }}
        defaultThemeName={DEFAULT_LIGHT_THEME_NAME}>
        <Table linesPerPage={1} data={mockData} headers={mockHeaders} />
      </ThemeProvider>
    );

    let line1column1 = queryByText("testl1c1");
    let line2column1 = queryByText("testl2c1");
    const back = queryByTestId("back");
    const next = queryByTestId("next");

    expect(line1column1).toBeTruthy();
    expect(line2column1).toBeFalsy();
    expect(back).toBeTruthy();
    expect(next).toBeTruthy();

    // nothing happens
    fireEvent.click(getByTestId("back"));
    line2column1 = queryByText("testl2c1");
    expect(line2column1).toBeFalsy();

    // next
    fireEvent.click(getByTestId("next"));
    line1column1 = queryByText("testl1c1");
    line2column1 = queryByText("testl2c1");
    expect(line2column1).toBeTruthy();
    expect(line1column1).toBeFalsy();

    // back
    fireEvent.click(getByTestId("back"));
    line1column1 = queryByText("testl1c1");
    line2column1 = queryByText("testl2c1");
    expect(line2column1).toBeFalsy();
    expect(line1column1).toBeTruthy();
  });
});
