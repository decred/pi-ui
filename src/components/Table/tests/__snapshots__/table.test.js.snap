// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component Matches the snapshot 1`] = `
<div
  className="tableWrapper"
>
  <table
    className="table"
  >
    <thead
      className="tableHead"
    >
      <tr
        className="tableRow"
      >
        <th
          className="tableHeadCell"
          scope="col"
        >
          Test1
        </th>
        <th
          className="tableHeadCell"
          scope="col"
        >
          Test2
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="tableRow"
      >
        <td
          className="tableBodyCell"
          data-label="Test1"
        >
          <span
            className="tableBodyCellText"
          >
            testl1c1
          </span>
        </td>
        <td
          className="tableBodyCell"
          data-label="Test2"
        >
          <span
            className="tableBodyCellText"
          >
            testl1c2
          </span>
        </td>
      </tr>
      <tr
        className="tableRow"
      >
        <td
          className="tableBodyCell"
          data-label="Test1"
        >
          <span
            className="tableBodyCellText"
          >
            testl2c1
          </span>
        </td>
        <td
          className="tableBodyCell"
          data-label="Test2"
        >
          <span
            className="tableBodyCellText"
          >
            testl2c2
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
